### PR TITLE
[ML] Upgrade to Pytorch 2.1.2 and zlib 1.2.13

### LINKED ISSUE #2587: [ML] Upgrade to PyTorch 2.1
PyTorch 2.1 shipped a couple of weeks ago, and we should upgrade to it providing it does not introduce significant backwards compatibility issues.

Our last upgrade was to PyTorch 1.13.1, and #2430 should provide inspiration for how to upgrade to PyTorch 2.1.

PyTorch 2.1 supports Python 3.11 as a beta feature. However, it also still supports Python 3.10. We should be able to continue to use Python 3.10 during the PyTorch build process to minimise the amount of build system changes and hence expedite the upgrade.

### PR DESCRIPTION
**Pull request by edsavage**
Pytorch 2.1.2 has recently been released. Opening this PR to upgrade to it and potentially flush out any BWC issues via the unit and ES integration tests.

Closes #2587

### CODE REVIEW

File: 3rd_party/licenses/pytorch-INFO.csv
Comment (droberts195) [inline 3rd_party/licenses/pytorch-INFO.csv:2]: The Git commit (3rd value) should also be changed here.

File: build-setup/linux.md
Comment (droberts195) [inline build-setup/linux.md:405]:
```suggestion
Building IPEX requires a lot of memory. To reduce this requirement, we can set the `MAX_JOBS` environment variable to lower the number of parallel processes:
```

File: build-setup/linux.md
Comment (droberts195) [inline build-setup/linux.md:412]: This needs some due diligence. The library that gets installed here is `libblas.so.3`. Does IPEX end up requiring that at runtime? If so then that's a problem as we cannot guarantee that it will be present on end user systems.

Or if IPEX doesn't need to link `libblas` at runtime, why is it needed at build time? Are there environment variables that can be set to say not to use this?

File: build-setup/windows.md
Comment (droberts195) [inline build-setup/windows.md:76]: This means `3rd_party/licenses/zlib-INFO.csv` needs updating too.

File: dev-tools/docker/linux_image/Dockerfile
Comment (droberts195) [inline dev-tools/docker/linux_image/Dockerfile:22]: This list is no longer in alphabetical order.

But also, please investigate my other comment about whether this is really needed, if so why, and is it causing a runtime complication.

**Comment by droberts195** [comment]
We usually release-note major library upgrades like this, so please add a release note like https://github.com/elastic/ml-cpp/pull/2430/files#diff-6eabb97e98640429cb68b61823f99438d287d604d72ecf0e93ed2e9ecbec7479R35 and https://github.com/elastic/ml-cpp/pull/2253/files#diff-6eabb97e98640429cb68b61823f99438d287d604d72ecf0e93ed2e9ecbec7479R37.

File: build-setup/linux.md
Comment (droberts195) [inline build-setup/linux.md:410]: Please can you confirm with `ldd` that the `BLAS=0` workaround works. Also check if there are any other niche OS library dependencies of either IPEX or PyTorch 2.1.

I am a bit worried about the side effects of setting `BLAS=0` for `LIBXSMM`. It might mean that then certain operations of `LIBXSMM` simply don't work. We'd only detect this at runtime when a model that uses IPEX needs to do one of these operations. Our PR build tests are nowhere near up to the job of detecting subtle problems like this.

For 8.12 this doesn't actually matter, as we're not using IPEX. However, this is an extra piece of work to do before we ever do use in in the future.

This line in the `LIBXSMM` `Makefile` makes me wonder if it's possible to force the BLAS to always be MKL, and then we could set the build up like that: https://github.com/libxsmm/libxsmm/blob/54b8bb32042d204ff6fe550fa70d961f22c0b99e/Makefile#L246

Please can you add a TODO in this file to revisit `BLAS=0` before actually using IPEX in production, and also open an issue to investigate the best way to build IPEX that's both portable to all supported Linux distributions _and_ means that IPEX won't have any missing areas of functionality that cause runtime errors.

But don't try and solve this as part of this PR.

Comment (droberts195) [reply build-setup/linux.md:410]: Those dynamic dependencies you've listed look good.

The BLAS documentation for `LIBXSMM` isn't 100% clear to me, and then there's the extra complexity on top of when/how IPEX uses `LIBXSMM`.

Thanks for opening #2589. Nothing more is needed for this PR though.

**Comment by edsavage** [comment]
buildkite build this please and thank you

**Comment by edsavage** [comment]
buildkite build this

**Comment by droberts195** [comment]
Looks good, but we need to run the QA suite against these changes before merging to `main`, as `main` could get promoted into serverless production immediately and we don't want to find out there are problems after this.

I think some manual compatibility testing with the current version of Eland should also be done before merging this. So build an Elasticsearch locally that incorporates a C++ build from this PR branch, then use the current version of Eland (which uses PyTorch 1.13.1) to download a model from HuggingFace, install it, and start it. Then perform at least one inference against that model. (The latest main branch snapshot of Kibana should be able to connect to the locally built Elasticsearch for this testing.)

Finally, with that same local install of Kibana and Elasticsearch, check that you can download and install ELSER v2 and do an inference using that.

**Comment by edsavage** [comment]
> manual compatibility testing

Sanity checks were performed using `elastic/distilbert-base-uncased-finetuned-conll03-english` for named entity recognition and `typeform/distilbert-base-uncased-mnli` for zero shot classification. Both of these models were imported  from `huggingface` via the `eland_import_hub_model` script. No issues were encountered in either of these tasks.

Likewise, no issues were encountered downloading ELSER v2 and testing against it.

**Comment by droberts195** [comment]
TODO: Set `QAF_TESTS_TO_RUN` to the appropriate value and then trigger the QA framework CI job.

**Comment by droberts195** [comment]
The crashes we've observed with the Intel-optimised variant of ELSER are caused by IPEX. I have opened https://github.com/intel/intel-extension-for-pytorch/issues/484 for this. Depending on the outcome of that we will either have to upgrade to a patched version of IPEX in this PR, or remove it entirely.

File: docs/CHANGELOG.asciidoc
Comment (droberts195) [inline docs/CHANGELOG.asciidoc:47]: Please delete these two lines. They're duplicated about 10 lines below in the 8.11.2 section, and 8.11.2 seems to be correct. (I think this might be cause by one of my attempts to resolve a merge clash - sorry about that.)

**Review by droberts195 — APPROVED** [review]
LGTM if CI passes and if you could just adjust the release notes before merging - it's probably worth waiting for the current CI to finish before adjusting the release notes to expedite smoking out any other problems, then merge first thing tomorrow if all is good.

**Pull request merged**
